### ISSUE TITLE
Liger-Kernel fails to compute loss when using train+eval and SP>1

### ISSUE THREAD

**Issue by sfc-gh-sbekman** [snowflakedb/ArcticTraining#266]
note: This issue at the moment is FYI to compile my observations and what I have tried to make it work. I don't personally have a need to do eval while training so unless someone asks this might not be an issue for anybody so why bother as this is not our code failing.

----------------------

Liger-Kernel fails to compute loss when using train+eval and returns `None` - removing `eval` (leaving pure training) removes the problem. Tested with SFT trainer and SP>1.

Last I checked its fused_linear_cross_entropy feature was computing `backward` in `forward` w/o checking if the model is running under no_grad. So I tried to disable just that kernel with:

```
        return AutoLigerKernelForCausalLM.from_pretrained(
            self.config.name_or_path,
[...]
            fused_linear_cross_entropy=False,
        )
```
but it still returned `None`.

So for now I have added an exception:

```
        if self.config.model.type == "liger":
            # letting liger do fused logits+loss calculation
            outputs = self.model(**batch, use_cache=False)
            loss = outputs.loss
            if loss is None:
                raise ValueError("Liger-Kernel failed to compute loss (returned None) - it's known to fail with eval enabled along train steps.")
```

I also checked `shift_labels` contain valid non `-100` tokens.

if users want this feature we would need to escalate this with the Liger-Kernel devs.

**Comment by Tcc0403** [comment]
Hi, I would like to look into this issue. Do I just run this recipe [sequence-parallelism/run-h100-sp8-llama-8b-no-ulysses-yes-liger-no-extras.yml](https://github.com/snowflakedb/ArcticTraining/blob/main/projects/sequence-parallelism/run-h100-sp8-llama-8b-no-ulysses-yes-liger-no-extras.yml) to reproduce the issue? Or how do I enable `eval` mode with ArticTraining? Feel free to point me in some docs that could be helpful. I'm new to this project but eager to help!

**Comment by sfc-gh-sbekman** [comment]
Thank you for reporting, @Tcc0403  - the size will be fixed here https://github.com/snowflakedb/ArcticTraining/pull/282

Let me create a recipe for the liger issue - the eval was added after ALST was released.

**Comment by sfc-gh-sbekman** [comment]
To reproduce:
```
# run-h200-sp2-llama-8b-liger-eval.yml
type: sft
micro_batch_size: 1
exit_iteration: 3
min_iterations: 3
sequence_parallel_size: 2

deepspeed:
  zero_optimization:
    stage: 3
    offload_optimizer:
      device: cpu
  seq_parallel_communication_data_type: bf16

optimizer:
  type: cpu_adam
  learning_rate: 0.00001

model:
  type: "liger"
  #name_or_path: meta-llama/Llama-3.1-8B-Instruct
  name_or_path: TinyLlama/TinyLlama-1.1B-Chat-v1.0
  attn_implementation: sdpa

data:
  train_eval_split: [0.8, 0.2]
  sources:
    - HuggingFaceH4/ultrachat_200k:train[:100]
  cache_dir: data-cache
  dl_num_workers: 1

  max_length: 16_000

eval_interval: 1

logger:
  level: WARNING
```

then prepare:

```
git clone https://github.com/snowflakedb/ArcticTraining && cd ArcticTraining
pip install liger-kernel
pip install -e .
```

then run with the yaml file from the top - I changed the model to a smaller one so it's much faster.

```
$ arctic_training run-h200-sp2-llama-8b-liger-eval.yml --master_port 9818
```

I rechecked it works fine with sp=1, so it must be something that is different when sp>1.

1. The first suspect code is:

https://github.com/snowflakedb/ArcticTraining/blob/8f61c507000d01b8cf30b4b13cad1dbc2945fde3/arctic_training/trainer/sft_trainer.py#L48-L95

but there is no difference between:

https://github.com/snowflakedb/ArcticTraining/blob/8f61c507000d01b8cf30b4b13cad1dbc2945fde3/arctic_training/trainer/sft_trainer.py#L48-L54

and:

https://github.com/snowflakedb/ArcticTraining/blob/8f61c507000d01b8cf30b4b13cad1dbc2945fde3/arctic_training/trainer/sft_trainer.py#L86C1-L89C1

the only reason I have a special branch for sp>1+liger is that I need to aggregate loss across ranks, but otherwise the code is exactly the same.

Practically the batch liger-kernels sees is divided by SP.

2. The second is the DL wrapper

https://github.com/snowflakedb/ArcticTraining/blob/8f61c507000d01b8cf30b4b13cad1dbc2945fde3/arctic_training/trainer/trainer.py#L246-L262

The data is sharded and computed a chunk on each rank.

Perhaps I missed something else, but training works just fine, somehow eval breaks this.

Thank you for looking into it, @Tcc0403

**Comment by sfc-gh-sbekman** [comment]
Aha! Thank you for discovering the issue and the fix, @Tcc0403!
